plan.go

```diff
@@ -0,0 +1,172 @@
+	"net/url"
+	"time"
+)
+
+// PlanStatus represents a plan state.
+type PlanStatus string
+
+//List all available plan statuses.
+const (
+	PlanCanceled   PlanStatus = "canceled"
+	PlanCreated    PlanStatus = "created"
+	PlanErrored    PlanStatus = "errored"
+	PlanFinished   PlanStatus = "finished"
+	PlanMFAWaiting PlanStatus = "mfa_waiting"
+	PlanPending    PlanStatus = "pending"
+	PlanQueued     PlanStatus = "queued"
+	PlanRunning    PlanStatus = "running"
+)
+
+// Plans handles communication with the plan related methods of the Terraform
+// Enterprise API.
+//
+// TFE API docs: https://www.terraform.io/docs/enterprise/api/plan.html
+type Plans struct {
+	client *Client
+}
+
+// Plan represents a Terraform Enterprise plan.
+type Plan struct {
+	ID               string                `jsonapi:"primary,plans"`
+	HasChanges       bool                  `jsonapi:"attr,has-changes"`
+	LogReadURL       string                `jsonapi:"attr,log-read-url"`
+	Status           PlanStatus            `jsonapi:"attr,status"`
+	StatusTimestamps *PlanStatusTimestamps `jsonapi:"attr,status-timestamps"`
+}
+
+// PlanStatusTimestamps holds the timestamps for individual plan statuses.
+type PlanStatusTimestamps struct {
+	CanceledAt   time.Time `json:"canceled-at"`
+	CreatedAt    time.Time `json:"created-at"`
+	ErroredAt    time.Time `json:"errored-at"`
+	FinishedAt   time.Time `json:"finished-at"`
+	MFAWaitingAt time.Time `json:"mfa_waiting-at"`
+	PendingAt    time.Time `json:"pending-at"`
+	QueuedAt     time.Time `json:"queued-at"`
+	RunningAt    time.Time `json:"running-at"`
+}
+
+// Read a single run by its ID.
+func (s *Plans) Read(ctx context.Context, planID string) (*Plan, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	u := fmt.Sprintf("plans/%s", url.QueryEscape(planID))
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Plan{}
+	err = s.client.do(ctx, req, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// Logs retrieves the logs of a plan.
+func (s *Plans) Logs(ctx context.Context, planID string) (io.Reader, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	// Get the plan to make sure it exists.
+	p, err := s.Read(ctx, planID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return an error if the log URL is empty.
+	if p.LogReadURL == "" {
+		return nil, fmt.Errorf("Plan %s does not have a log URL", planID)
+	}
+
+	return &LogReader{
+		client: s.client,
+		ctx:    ctx,
+		planID: p.ID,
+	}, nil
+}
+
+// LogReader implements io.Reader for streaming plan logs.
+type LogReader struct {
+	client *Client
+	ctx    context.Context
+	offset int64
+	planID string
+}
+
+func (r *LogReader) Read(l []byte) (int, error) {
+	for {
+		select {
+		case <-r.ctx.Done():
+			return 0, r.ctx.Err()
+		case <-time.After(500 * time.Millisecond):
+			p, err := r.client.Plans.Read(r.ctx, r.planID)
```

Comment:
To me this feels like something we should probably move out of the "hot path" - this adds significant time to the request and load on TFE to check every time we request the logs from Archivist. I think we could do something better like track the number of requests and just check on every 3rd or 5th one or so.

Also, we should move this below the actual request for the logs, because if that fails then we won't use this result anyways.